### ISSUE TITLE
Soundfield Recorder not working properly

### ISSUE THREAD

**Issue by mstergtr** [googleprivate/resonance-audio-unity-sdk#28]
When I use the soundfield recorder in the Resonance Audio Listener script only channels 1 and 4 contain audio (2 and 3 are silence). Sounds are all around and everything seems to be working properly otherwise.

On Unity 2017.3.1p4 and have the latest resonance SDK in my assets. 

UPDATE:  I was using timeline to trigger animations, which remain static when the resonance recorder is activated. A friend was able to edit the code to enable the sound field recorder to work during runtime and everything is recording properly now (although it does act a bit strange). I can share the code if others are interested.

**Comment by mstergtr** [comment]
Could someone verify that this is indeed a bug and not something unique to my setup? Maybe I need an audio interface that supports multichannel audio and set it to quad in the audio settings? Thanks

**Comment by pan-athen** [comment]
I think that the Google Resonance soundfield recorder that renders the sound setup within Unity, outputs audio in the Ambisonics format. That means that this is an encoded format, not ready to be played through a common quadraphonic setup. It's not quadraphonic, it's 360 information, stored in 4 channels of encoded information. Did you tried playing back your renders on an Ambisonic player that supports the format that Unity/Google Resonance outputs? I think that the latest VLC supports it but I'm not sure.

**Comment by mstergtr** [comment]
Yes, I know that it is not quad audio, but it is an encoded format contained in four channels. I have tried to playback the audio within Reaper using both the ambix headphone decoder and the resonance audio monitor. The problem (stated in my original post) is that channels 2 and 3 are completely silent, so there is no spatial information in the file. Another spatial audio capture tool I've used (VR panoramic) still requires the use of a four channel interface and the sound settings set to quad, making me think that same setup could be required for resonance. I'm hoping someone can verify that this is indeed a bug and not something I am doing wrong.

**Comment by pan-athen** [comment]
Oh I see.

Strange behavior indeed.

Sorry can't verify at the moment as I have a very complicated setup on my production machine.

Hope you find the solution. :)

**Comment by furlerer** [comment]
I can confirm I am currently experiencing the same issue.

Exported soundfields only have audio in channels 1 and 4, channels 2 and 3 are silent. 
I have tried different combinations of settings in the audio panel, but to no avail.

--------------------------------------------------------------------------------------------------

UPDATE: Although this issue seemed to be present in all versions of 2018.1 I used, I am happy to say since updating the project to 2018.2.0b7, my case appears to be fixed. Correct audio is now in all 4 channels.

**Comment by AnnaMy82** [comment]
Hi @furlerer and @mstergtr,

Just to confirm - are you exporting (rendering) a linear ambisonic audio mix from Unity? That's exactly what I need! I'd love to see the code, if you're happy to share? I presume Unity still only support 1st order for now? 
I'm familiar with Resonance but haven't had to record a soundfield until now, so glad to see there's something like 'record' in the Listener. Many thanks!

**Comment by mstergtr** [comment]
I have some edited resonance scripts that make the soundfield recorder more usable: https://drive.google.com/open?id=1jiSnNRjO-Zwf18Y2Ui0QUfxO-AgcSX-r

Just replace the existing scripts with these and use F and G to start/stop the recording (or set up a toggle). Note that it can take a while for the ambisonic file to show up in the assets folder after stopping the recording.

**Comment by MJRiordan** [comment]
Hi @mstergtr,

I'm working on a project for which it would be great to be able to use the sound field recorder at runtime, but when I tried using those scripts you shared, I still get the message saying it can only be used in edit mode. Is that the same code you were talking about in the update to your first post?

**Comment by mstergtr** [comment]
@MJRiordan I double checked the link and those are the correct scripts. I'm not sure why you would still be getting the message about edit mode.

**Comment by MJRiordan** [comment]
@mstergtr thanks for getting back to me! I eventually noticed there were a few lines about 'f' and 'g' (pasted below) commented out so I removed the comment slashes and it seems to be working now. Was I supposed to remove all comment slashes or just the ones surrounding these lines?

```
if (Input.GetKeyDown("f")) {
            IsRecording = false;
            recorderStartTime = 0.0;
            recorderTaggedSources = new List<AudioSource>();
            StartSoundfieldRecorder();
        }
        if (Input.GetKeyDown("g")) {
            string myFileName = string.Format("Assets/record-{0:yyyy-MM-dd_hh-mm-ss-tt}.ogg", System.DateTime.Now);
            Debug.Log(myFileName);
            StopSoundfieldRecorder(myFileName);
        }
```

**Comment by JonOlive** [comment]
@MJRiordan I dare say you'll have figured it out by now - but in case not - I suspect that you uncommented lines 161-164 which throw up the "can only be used in edit mode" message and quit the StartSoundifeldRecorder method

**Comment by KawhiZhao** [comment]
I notice that the soundField Recorder can only record in Unity editor mode not in player mode. Can I record the audio in player mode?

**Comment by MJRiordan** [comment]
> @MJRiordan I dare say you'll have figured it out by now - but in case not - I suspect that you uncommented lines 161-164 which throw up the "can only be used in edit mode" message and quit the StartSoundifeldRecorder method

@JonOlive I'm afraid it's been so long that I can't remember if that's what I did or not, but I think we did get it working eventually. I think I would have noticed that I wanted those lines commented out, but we were all tired and short on time at that point in that project, it's quite possible I didn't read it correctly.
